rclcpp/src/rclcpp/executor.cpp

```diff
@@ -741,6 +748,7 @@ Executor::wait_for_work(std::chrono::nanoseconds timeout)
 
   // check the null handles in the wait set and remove them from the handles in memory strategy
   // for callback-based entities
+  std::lock_guard<std::mutex> guard(mutex_);
   memory_strategy_->remove_null_handles(&wait_set_);
```

Comment:
It's possible that `memory_strategy_` is modified during `rcl_wait()`. Should we be concerned with that? I don't know enough about the executor implementation to be sure.

Reply:
> Should we be concerned with that?

`rcl_wait()` cannot modify `memory_strategy_`.
The reason to not hold `mutex_` locked while calling `rcl_wait()` is that in the following situation:
- Thread 1: `executor.spin();`
- Thread 2: `executor.add_node(node);`

Thread 2 might wait long if there happens to not be any event awakening the executor.

Reply:
But while thread 1 is spinning, thread 2 can modify `memory_strategy_`, right? And then we are using the result of `rcl_wait()` to update `memory_strategy_` here.

Reply:
Good question, yes but that's ok.

`remove_node()`, `add_node()` only add/remove a `guard_condition` to `memory_strategy_`, and `remove_null_handles()`, `get_next_*()` methods aren't affected by that change.

There's a problem if you add or remove other entities (i.e. not a guard condition, e.g. a subscription) to `memory_strategy_` in the middle, but that only happens at the start of `wait_for_work()` with the mutex taken.